src/mono/mono/eventpipe/ds-ipc.c

```diff
@@ -186,115 +185,133 @@ void
 ipc_log_poll_handles (ds_rt_ipc_poll_handle_array_t *ipc_poll_handles)
 {
 	// TODO: Should this be debug only?
-	ds_rt_ipc_poll_handle_array_iterator_t ipc_poll_handles_iterator;
 	DiagnosticsIpcPollHandle ipc_poll_handle;
 	ep_char8_t buffer [DS_IPC_MAX_TO_STRING_LEN];
 	uint32_t connection_id = 0;
 
-	ds_rt_ipc_poll_handle_array_iterator_begin (ipc_poll_handles, &ipc_poll_handles_iterator);
+	ds_rt_ipc_poll_handle_array_iterator_t ipc_poll_handles_iterator = ds_rt_ipc_poll_handle_array_iterator_begin (ipc_poll_handles);
 	while (!ds_rt_ipc_poll_handle_array_iterator_end (ipc_poll_handles, &ipc_poll_handles_iterator)) {
 		ipc_poll_handle = ds_rt_ipc_poll_handle_array_iterator_value (&ipc_poll_handles_iterator);
 		if (ipc_poll_handle.ipc) {
 			if (!(ds_ipc_to_string (ipc_poll_handle.ipc, buffer, EP_ARRAY_SIZE (buffer)) > 0))
 				buffer [0] = '\0';
 			DS_LOG_INFO_2 ("\tSERVER IpcPollHandle[%d] = %s\n", connection_id, buffer);
 		} else {
-			if (!(ds_ipc_stream_to_string (ipc_poll_handle.stream, buffer, EP_ARRAY_SIZE (buffer)) > 0))
+			if (!(ds_ipc_stream_to_string (ipc_poll_handle.stream, buffer, EP_ARRAY_SIZE (buffer))))
 				buffer [0] = '\0';
 			DS_LOG_INFO_2 ("\tCLIENT IpcPollHandle[%d] = %s\n", connection_id, buffer);
 		}
-		ds_rt_ipc_poll_handle_array_iterator_next (ipc_poll_handles, &ipc_poll_handles_iterator);
+		ds_rt_ipc_poll_handle_array_iterator_next (&ipc_poll_handles_iterator);
 		connection_id++;
 	}
 }
 
-void
+bool
 ds_ipc_stream_factory_init (void)
 {
 	ds_rt_port_array_alloc (&_ds_port_array);
+	return ds_rt_port_array_is_valid (&_ds_port_array);
 }
 
 void
 ds_ipc_stream_factory_fini (void)
 {
-	ds_rt_port_array_free (&_ds_port_array);
+	// TODO: Race between server thread and shutdown, _ds_port_array and ports can not be freed without resolving
+	// that race first. Diagnostic server thread is currently designed to not break waits on
+	// shutdown unless clients activity wakes server thread.
```

Comment:
Is this race specific to the C implementation, or shared with the C++ diagnostic server, too?

Reply:
Its in both, but CoreClr don't always do clean shutdown, so might not depend on thread actually quitting, so not a problem on their end, they don't do any cleanup regarding the ports either.